pallets/liquidity-pools-forwarder/src/tests.rs

```diff
@@ -0,0 +1,393 @@
+		/// panics in mock because `Message::NonForward` serialization is
+		/// expected
+		fn with_missing_forward_info() {
+			System::externalities().execute_with(|| {
+				config_mocks(Message::Forward, false);
+
+				assert_noop!(
+					<LiquidityPoolsForwarder as MessageSender>::send(
+						ROUTER_ID,
+						FORWARDER_DOMAIN_ADDRESS,
+						Message::Forward
+					),
+					Error::<Runtime>::ForwardInfoNotFound
+				);
+			});
+		}
+
+		#[test]
+		#[should_panic]
+		/// Attempting to send forwarded message panics here in mock
+		/// because `Message::NonForward` serialization is expected
+		fn with_expected_non_forward_serialization() {
+			System::externalities().execute_with(|| {
+				config_mocks(Message::NonForward, true);
+
+				assert_ok!(<LiquidityPoolsForwarder as MessageSender>::send(
+					ROUTER_ID,
+					FORWARDER_DOMAIN_ADDRESS,
+					Message::NonForward
+				));
+			});
+		}
+
+		#[test]
+		#[should_panic]
+		/// Attempting to send non-forwarded message panics here in mock
+		/// because `Message::Forward` serialization is expected
+		fn with_expected_forward_serialization() {
+			System::externalities().execute_with(|| {
+				config_mocks(Message::Forward, false);
+
+				assert_ok!(<LiquidityPoolsForwarder as MessageSender>::send(
+					ROUTER_ID,
+					FORWARDER_DOMAIN_ADDRESS,
+					Message::NonForward
+				));
+			});
+		}
+
+		#[test]
+		fn with_nesting() {
+			System::externalities().execute_with(|| {
+				config_mocks(Message::Forward, true);
+
+				assert_noop!(
+					<LiquidityPoolsForwarder as MessageSender>::send(
+						ROUTER_ID,
+						FORWARDER_DOMAIN_ADDRESS,
+						Message::Forward
+					),
+					ERROR_NESTING
+				);
+			});
+		}
+
+		#[test]
+		fn non_forward_with_message_receiver_err() {
+			System::externalities().execute_with(|| {
+				config_mocks(Message::Forward, true);
+				MockSenderReceiver::mock_send(|_, _, _| Err(ERROR));
+
+				assert_noop!(
+					<LiquidityPoolsForwarder as MessageSender>::send(
+						ROUTER_ID,
+						FORWARDER_DOMAIN_ADDRESS,
+						Message::NonForward
+					),
+					ERROR
+				);
+			});
+		}
+	}
+}
+
+mod receive_message {
+	use cfg_traits::liquidity_pools::MessageReceiver;
+
+	use super::*;
+
+	fn config_mocks(set_forwarding_info: bool) {
+		MockSenderReceiver::mock_receive(move |middleware, origin, message| {
+			assert_eq!(middleware, ROUTER_ID);
+			assert_eq!(origin, FORWARDER_DOMAIN_ADDRESS);
+			assert_eq!(message, Message::NonForward);
+			Ok(())
+		});
+
+		if set_forwarding_info {
+			assert_ok!(LiquidityPoolsForwarder::set_forwarder(
+				RuntimeOrigin::root(),
+				ROUTER_ID,
+				SOURCE_DOMAIN,
+				FORWARD_CONTRACT
+			));
+		}
+	}
+
+	mod success {
+		use cfg_traits::liquidity_pools::MessageReceiver;
+
+		use super::*;
+
+		#[test]
+		fn with_forwarding() {
+			System::externalities().execute_with(|| {
+				config_mocks(true);
+
+				assert_ok!(<LiquidityPoolsForwarder as MessageReceiver>::receive(
```

Comment:
Just a NIT curiosity, could not be `LiquidityPoolsForwarder::receive()`?

Reply:
Sure but I wanted to make it explicit.

Reply:
[4a0f751](https://github.com/centrifuge/centrifuge-chain/pull/1971/commits/4a0f751c6638212341ca1b344c5832e0f6a425f9)